jetty-core/jetty-osgi/src/main/java/org/eclipse/jetty/osgi/OSGiDeployableBundleMetadata.java

```diff
@@ -115,139 +130,90 @@ private static String getContextPath(Bundle bundle)
+    public static void registerAsOSGiService(ContextHandler contextHandler)
+    {
+        ServiceRegistration<?> serviceRegistration = (ServiceRegistration<?>)contextHandler.getAttribute(REGISTRATION);
+        if (serviceRegistration == null)
         {
-            String key = keys.nextElement();
-            String val = headers.get(key);
-            if (Deployable.ENVIRONMENT.equalsIgnoreCase(key) || OSGiWebappConstants.JETTY_ENVIRONMENT.equalsIgnoreCase(key))
-                getProperties().put(Deployable.ENVIRONMENT, val);
-            else if (Deployable.DEFAULTS_DESCRIPTOR.equalsIgnoreCase(key) || OSGiWebappConstants.JETTY_DEFAULT_WEB_XML_PATH.equalsIgnoreCase(key))
-            {
-                getProperties().put(Deployable.DEFAULTS_DESCRIPTOR, val);
-            }
-            else if (OSGiWebappConstants.JETTY_WEB_XML_PATH.equalsIgnoreCase(key))
-            {
-                getProperties().put(key, val);
-            }
-            else if (OSGiWebappConstants.JETTY_CONTEXT_FILE_PATH.equalsIgnoreCase(key))
-            {
-                getProperties().put(key, val);
-            }
-        }
+            Dictionary<String, String> properties = new Hashtable<>();
+            properties.put(OSGiWebappConstants.WATERMARK, OSGiWebappConstants.WATERMARK);
 
-        //set up the context path based on the supplied value, or the calculated default
-        setContextPath(getContextPath(bundle));
-    }
+            String bundleSymbolicName = getBundleSymbolicName(contextHandler);
+            if (StringUtil.isNotBlank(bundleSymbolicName))
+                properties.put(OSGiWebappConstants.OSGI_WEB_SYMBOLICNAME, bundleSymbolicName);
 
-    public Resource getBundleResource()
-    {
-        return _bundleResource;
-    }
+            String bundleVersion = getBundleVersionAsString(contextHandler);
+            if (StringUtil.isNotBlank(bundleVersion))
+                properties.put(OSGiWebappConstants.OSGI_WEB_VERSION, bundleVersion);
 
-    @Override
-    public ContextHandler getContextHandler() throws Exception
-    {
-        if (_contextHandler == null)
-                _contextHandler = getAppProvider().createContextHandler(this);
-            return _contextHandler;
-    }
-
-    public void setContextHandler(ContextHandler contextHandler)
-    {
-        _contextHandler = contextHandler;
-    }
+            properties.put(OSGiWebappConstants.OSGI_WEB_CONTEXTPATH, contextHandler.getContextPath());
 
-    public String getPathToResourceBase()
-    {
-        return _pathToResourceBase;
+            serviceRegistration = FrameworkUtil.getBundle(OSGiDeployableBundleMetadata.class).getBundleContext().registerService(ContextHandler.class.getName(), contextHandler, properties);
+            contextHandler.setAttribute(REGISTRATION, serviceRegistration);
+        }
     }
 
-    public void setPathToResourceBase(String path)
+    public static void setBundle(ContextHandler contextHandler, Bundle bundle)
```

Comment:
I don't think this is exactly the right place to do this, can be done by either the `ContextFactory` impl, or the `AbstractContextProvider`.

Reply:
How about moving all of the OSGI specific ContextHandler statics that use/need/manipulate the attributes in a ContextHandler to a new `OSGiContextHandlerUtil` class?

Reply:
Done